pkg/controllers/management/cis/scheduled_scan.go

```diff
@@ -0,0 +1,266 @@
+	clusterLister v3.ClusterLister,
+	clusterScanClient v3.ClusterScanInterface,
+	clusterScanLister v3.ClusterScanLister,
+) *scheduledScanHandler {
+	return &scheduledScanHandler{
+		managementClient:  managementClient,
+		clusterClient: clusterClient,
+		clusterLister:     clusterLister,
+		clusterScanClient: clusterScanClient,
+		clusterScanLister: clusterScanLister,
+		clustersMap:       make(map[string]*scheduleInfo),
+	}
+}
+
+func (i *scheduleInfo) String() string {
+	return fmt.Sprintf("CronSchedule: %v lastAt: %v nextAt: %v",
+		i.CronSchedule, i.lastScanLaunchedAt.String(), i.nextScanAt.String())
+}
+
+func (ssh *scheduledScanHandler) clusterSync(key string, cluster *v3.Cluster) (runtime.Object, error) {
+	if cluster == nil ||
+		cluster.DeletionTimestamp != nil ||
+		cluster.Spec.ScheduledClusterScan == nil {
+		return nil, nil
+	}
+
+	clusterInfo, ok := ssh.clustersMap[cluster.Name]
+	if cluster.Spec.ScheduledClusterScan.Enabled {
+		if !ok {
+			// Do stuff to enable scheduling
+			cronSchedule := getCronSchedule(cluster)
+			clusterInfo = &scheduleInfo{
+				CronSchedule: cronSchedule,
+			}
+			logrus.Infof("scheduledScanHandler: clusterSync: Scheduled Scan enabled for cluster %v: %v",
+				cluster.Name, cronSchedule)
+			if err := scheduleScan(cluster, ssh.clusterScanClient, clusterInfo); err != nil {
+				return nil, fmt.Errorf("scheduledScanHandler: clusterSync: error scheduling scan: %v", err)
+			}
+			ssh.clustersMap[cluster.Name] = clusterInfo
+		} else {
+			if cluster.Spec.ScheduledClusterScan.ScheduleConfig != nil &&
+				cluster.Spec.ScheduledClusterScan.ScheduleConfig.CronSchedule != "" &&
+				cluster.Spec.ScheduledClusterScan.ScheduleConfig.CronSchedule != clusterInfo.CronSchedule {
+				clusterInfo.CronSchedule = cluster.Spec.ScheduledClusterScan.ScheduleConfig.CronSchedule
+				logrus.Infof("scheduledScanHandler: clusterSync: cron schedule modified for cluster %v: %v",
+					cluster.Name, clusterInfo.CronSchedule)
+			}
+		}
+	} else {
+		if ok {
+			logrus.Infof("scheduledScanHandler: clusterSync: Scheduled Scan disabled for cluster: %v",
+				cluster.Name)
+			delete(ssh.clustersMap, cluster.Name)
+		}
+	}
+	return nil, nil
+}
+
+func (ssh *scheduledScanHandler) sync(key string, _ *v3.ClusterScan) (runtime.Object, error) {
+	if !strings.HasPrefix(key, ScheduledScanKeyPrefix) {
+		return nil, nil
+	}
+	splits := strings.Split(strings.TrimPrefix(key, ScheduledScanKeyPrefix), separator)
+	clusterId := splits[0]
+	pastCronSchedule := splits[1]
+	clusterInfo, ok := ssh.clustersMap[clusterId]
+	if !ok {
+		logrus.Infof("scheduledScanHandler: sync: scheduled scan has been disabled, ignoring")
+	}
+	cluster, err := ssh.clusterLister.Get("", clusterId)
+	if err != nil {
+		return nil, fmt.Errorf("scheduledScanHandler: sync: error fetching cluster %v: %v", clusterId, err)
+	}
+
+	if !cluster.Spec.ScheduledClusterScan.Enabled {
+		logrus.Infof("scheduledScanHandler: sync: scheduled scan is disabled, ignoring the current one")
```

Comment:
wondering if it'd be ok to delete this from clustersMap?

Reply:
I have the delete logic in the clustersync when the schedule is disabled, not necessary to delete it here. It's just cleaner to have it in one place.